File: packages/governance/rust/src/tally.rs
use serde::Deserialize;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::ids::compute_tally_hash;
use crate::types::{GovernanceConfig, Proposal, ProposalStatus, Vote, VoteTally};

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

#[derive(Deserialize)]
pub struct TallyVotesParams {
    pub proposal: Proposal,
    pub votes: Vec<Vote>,
    pub total_weight: f64,
    pub quorum_weight: Option<f64>,
    pub config: Option<GovernanceConfig>,
}

pub fn tally_votes(params: TallyVotesParams) -> Result<VoteTally, String> {
    if params.votes.is_empty() {
        return Err("no votes to tally".to_string());
    }

    let now = now_ms();
    if now < params.proposal.voting_ends_at {
        return Err("voting has not ended yet".to_string());
    }

    let algorithm = match params.config {
        Some(ref c) if c.voting.algorithm == "quadratic" => "quadratic",
        _ => "linear",
    };

    let mut yes_weight = 0.0_f64;
    let mut no_weight = 0.0_f64;
    let mut abstain_weight = 0.0_f64;

    if algorithm == "quadratic" {
        for vote in &params.votes {
            let sqrt_weight = if vote.weight > 0.0 {
                vote.weight.sqrt()
            } else {
                0.0
            };
            match vote.choice.as_str() {
                "yes" => yes_weight += sqrt_weight,
                "no" => no_weight += sqrt_weight,
                "abstain" => abstain_weight += sqrt_weight,
                _ => {}
            }
        }
    } else {
        for vote in &params.votes {
            match vote.choice.as_str() {
                "yes" => yes_weight += vote.weight,
                "no" => no_weight += vote.weight,
                "abstain" => abstain_weight += vote.weight,
                _ => {}
            }
        }
    }

    let total_cast = yes_weight + no_weight + abstain_weight;

    let quorum_weight = params.quorum_weight.unwrap_or_else(|| {
        if let Some(ref config) = params.config {
            ((params.total_weight * config.voting.quorum_bps as f64) / 10000.0).floor()
        } else {
            params.total_weight
        }
    });

    let quorum_reached = total_cast >= quorum_weight;

    let pass_threshold_bps = params
        .config
        .as_ref()
        .map(|c| c.voting.pass_threshold_bps)
        .unwrap_or(5000);

    let passed = if quorum_reached && (yes_weight + no_weight) > 0.0 {
        let ratio = (yes_weight / (yes_weight + no_weight)) * 10000.0;
        ratio >= pass_threshold_bps as f64
    } else {
        false
    };

    let floor_3 = |v: f64| (v * 1000.0).floor() / 1000.0;

    let tally = VoteTally {
        proposal_id: params.proposal.id,
        yes: floor_3(yes_weight),
        no: floor_3(no_weight),
        abstain: floor_3(abstain_weight),
        total_weight: params.total_weight,
        quorum_weight,
        quorum_reached,
        passed,
        threshold_bps: pass_threshold_bps,
        algorithm: algorithm.to_string(),
    };

    Ok(tally)
}

pub fn finalize_proposal(proposal: &Proposal, tally: &VoteTally) -> Proposal {
    let mut p = proposal.clone();
    p.status = if tally.passed {
        ProposalStatus::Passed
    } else {
        ProposalStatus::Failed
    };
    p.vote_tally = Some(tally.clone());
    p
}

pub fn compute_tally_proof_hash(tally: &VoteTally) -> String {
    compute_tally_hash(tally)
}

File: packages/governance/src/tally.ts
import type { Proposal, Vote, VoteTally, GovernanceConfig, GovernanceResult, MembershipSnapshot, Delegation } from './types.js'
import { computeTallyHash, computeVoteId } from './ids.js'
import { getMemberWeight, getTotalWeight, verifyMembershipSnapshot } from './snapshot.js'
import { getActiveDelegations } from './delegation.js'

export function tallyVotes(params: {
  proposal: Proposal
  votes: Vote[]
  snapshot: MembershipSnapshot
  config?: GovernanceConfig
  delegations?: Delegation[]
  now?: number
}): GovernanceResult<VoteTally> {
  const { proposal, votes, snapshot } = params

  if (votes.length === 0) {
    return { error: 'no votes to tally' }
  }

  if (proposal.status !== 'active') {
    return { error: `proposal is in status '${proposal.status}', cannot tally` }
  }

  if (snapshot.daoId !== proposal.daoId || snapshot.hash !== proposal.membershipSnapshotHash) {
    return { error: 'membership snapshot does not match proposal' }
  }
  if (!verifyMembershipSnapshot(snapshot)) {
    return { error: 'membership snapshot integrity check failed' }
  }

  const now = params.now ?? Date.now()
  if (now < proposal.votingEndsAt) {
    return { error: 'voting has not ended yet' }
  }

  const algorithm = params.config?.voting?.algorithm === 'quadratic' ? 'quadratic' : 'linear'
  const quadraticConfig = params.config?.voting?.quadratic

  const seenVoteIds = new Set<string>()
  const seenBallots = new Set<string>()
  const seenVoters = new Set<string>()
  const spentCreditsByVoter = new Map<string, number>()
  const totalWeight = getTotalWeight(snapshot, snapshot.frozenAt)

  let yesWeight = 0
  let noWeight = 0
  let abstainWeight = 0

  for (const vote of votes) {
    if (vote.proposalId !== proposal.id) {
      return { error: `vote ${vote.id} belongs to a different proposal` }
    }
    if (seenVoteIds.has(vote.id)) {
      return { error: `duplicate vote ID: ${vote.id}` }
    }
    seenVoteIds.add(vote.id)

    // Ballot ID binding: every vote's ID is deterministic given its contents,
    // so a forged or recomputed ballot cannot claim an arbitrary ID.
    const expectedVoteId = computeVoteId(vote.proposalId, vote.voter, vote.choice, vote.castAt)
    if (vote.id !== expectedVoteId) {
      return { error: `vote ID ${vote.id} does not match its ballot contents` }
    }

    const ballotKey = `${vote.voter}:${vote.choice}`
    if (seenBallots.has(ballotKey)) {
      return { error: `duplicate ballot for ${vote.voter} and ${vote.choice}` }
    }
    seenBallots.add(ballotKey)
    if (algorithm !== 'quadratic' && seenVoters.has(vote.voter)) {
      return { error: `duplicate ballot for voter ${vote.voter}` }
    }
    seenVoters.add(vote.voter)

    if (!['yes', 'no', 'abstain'].includes(vote.choice)) {
      return { error: `invalid vote choice for ${vote.voter}` }
    }
    if (vote.castAt < proposal.votingStartsAt || vote.castAt > proposal.votingEndsAt) {
      return { error: `vote ${vote.id} was cast outside the voting interval` }
    }

    const memberWeight = getMemberWeight(snapshot, vote.voter, snapshot.frozenAt)
    if (memberWeight <= 0) {
      return { error: `voter ${vote.voter} is not eligible in the membership snapshot` }
    }
    if (!Number.isFinite(vote.weight) || vote.weight <= 0 || vote.weight > memberWeight) {
      return { error: `invalid vote weight for ${vote.voter}` }
    }

    // Delegations must resolve to real, active delegation records when a vote
    // claims to have been cast on behalf of someone else.
    if (vote.delegationChain && vote.delegationChain.length > 0) {
      if (!params.delegations) {
        return { error: `vote ${vote.id} has a delegation chain but no delegation records were provided` }
      }
      const chainError = validateDelegatedBallot(vote, params.delegations, proposal.daoId, snapshot, now)
      if (chainError) return { error: chainError }
    } else if (algorithm !== 'quadratic' && vote.weight !== memberWeight) {
      return { error: `vote weight for ${vote.voter} does not match snapshot weight` }
    }

    if (algorithm === 'quadratic') {
      if (!Number.isSafeInteger(vote.weight) || vote.quadraticCredits !== vote.weight * vote.weight) {
        return { error: `invalid quadratic weight or credits for ${vote.voter}` }
      }
      // Aggregate budget: a voter must not exceed their total quadratic credit
      // budget across ALL of their ballots, not just within a single ballot.
      const spent = (spentCreditsByVoter.get(vote.voter) ?? 0) + (vote.quadraticCredits ?? 0)
      const budget =
        quadraticConfig?.creditSource === 'fixed'
          ? (quadraticConfig.maxCreditsPerMember ?? memberWeight)
          : memberWeight
      if (spent > budget) {
        return {
          error: `aggregate quadratic credits for ${vote.voter} (${spent}) exceed budget of ${budget}`,
        }
      }
      spentCreditsByVoter.set(vote.voter, spent)
    }

    if (vote.choice === 'yes') yesWeight += vote.weight
    else if (vote.choice === 'no') noWeight += vote.weight
    else abstainWeight += vote.weight
  }

  const totalCast = yesWeight + noWeight + abstainWeight
  const quorumWeight = params.config
    ? Math.floor((totalWeight * params.config.voting.quorumBps) / 10000)
    : totalWeight

  const quorumReached = totalCast >= quorumWeight
  const passThresholdBps = params.config?.voting?.passThresholdBps ?? 5000
  const decided = yesWeight + noWeight
  const passed = quorumReached && decided > 0 && (yesWeight / decided) * 10000 >= passThresholdBps

  const tally: VoteTally = {
    proposalId: proposal.id,
    yes: Math.floor(yesWeight * 1000) / 1000,
    no: Math.floor(noWeight * 1000) / 1000,
    abstain: Math.floor(abstainWeight * 1000) / 1000,
    totalWeight,
    quorumWeight,
    quorumReached,
    passed,
    thresholdBps: passThresholdBps,
    algorithm,
  }

  return tally
}

/**
 * Verify that a vote claiming to be delegated matches real delegation records:
 * - the chain must start at the voter who casts the ballot
 * - every hop must be an active delegation record for the proposal
 * - the weight must not exceed what the delegation actually transfers
 */
function validateDelegatedBallot(
  vote: Vote,
  delegations: Delegation[],
  daoId: string,
  snapshot: MembershipSnapshot,
  now: number,
): string | undefined {
  const chain = vote.delegationChain!
  if (chain[0] !== vote.voter) {
    return `delegation chain for ${vote.voter} does not start with the voter`
  }
  const active = getActiveDelegations(delegations, daoId, now)
  for (let i = 0; i < chain.length - 1; i++) {
    const hop = active.find((d) => d.delegator === chain[i] && d.delegate === chain[i + 1])
    if (!hop) {
      return `delegation ${chain[i]} -> ${chain[i + 1]} in vote ${vote.id} is not an active delegation record`
    }
  }
  const delegatorWeight = getMemberWeight(snapshot, vote.voter, snapshot.frozenAt)
  const firstHop = active.find((d) => d.delegator === chain[0] && d.delegate === chain[1])
  const transferred = firstHop && firstHop.weight > 0 ? Math.min(firstHop.weight, delegatorWeight) : delegatorWeight
  if (vote.weight > transferred) {
    return `delegated vote weight for ${vote.voter} exceeds delegated weight`
  }
  return undefined
}

export function finalizeProposal(
  proposal: Proposal,
  tally: VoteTally,
): Proposal {
  if (proposal.status !== 'active') {
    throw new Error(`cannot finalize proposal in status '${proposal.status}'`)
  }
  if (tally.proposalId !== proposal.id) {
    throw new Error('tally does not belong to proposal')
  }
  const status = tally.passed ? 'passed' : 'failed'
  return { ...proposal, status, voteTally: tally }
}

export function computeTallyProofHash(tally: VoteTally): string {
  return computeTallyHash(tally)
}
